guide/secure_secret_management.md

```diff
@@ -0,0 +1,165 @@
+
+OpenFaaS deploys functions as Docker Swarm Services, as result there are several features that we can leverage to simplify the development and subsquent deployment of functions to hardened production environments.
+
+## Using Environment Variables
+First, and most simple, is the ability to set environment variables at deploy time. For example, you might want to set the `NODE_ENV` or `DEBUG` variable.  Setting the `NODE_ENV` while using the `faas-cli` might look like this
+
+```sh
+$ faas-cli deploy -f ./samples.yml
+$ fass-cli invoke nodehelloenv
+Hello from a production machine
+```
+
+Where your `samples.yml` stack file looks like this
+```yaml
+provider:
+  name: faas
+  gateway: http://localhost:8080
+
+functions:
+  nodehelloenv:
+    lang: Dockerfile
+    skip_build: true
+    image: functions/nodehelloenv:latest
+    environment:
+        NODE_ENV: production
+```
+
+
+(_Optional_) If you are directly using the OpenFaaS Gateway API, then it will look like this
+
+```sh
+$ curl -H "Content-Type: application/json" \
+     -X POST \
+     -d '{"service":"nodeinfo","network":"func", "image": "functions/nodehelloenv:latest", "envVars": {"NODE_ENV": "production"}}' \
+     http://localhost:8080/system/functions
+$ curl -X POST \
+       -H 'Content-Type: text/plain' \
+       -H 'Content-Length: 0' \
+       http://localhost:8080/function/nodehelloenv
+Hello from a production machine
+```
+
+
+## Using Swarm Secrets
+A very tempting thing to do is to now add database password or api secrets as environment variables.  However, this is not secure.  Instead, we can leverage the [Docker Swarm Secrets](https://docs.docker.com/engine/swarm/secrets/) feature to safely store and give our functions access to the needed values. Using secrets is a two step process.  Take the [ApiKeyProtected](../sample-functions/ApiKeyProtected) example function, when we deploy this function we provide a secret key that it uses to authenticate requests to it.  First we must add a secret to the swarm
+
+```sh
+docker secret create secret_api_key ~/secrets/secret_api_key.txt
+```
+
+where `~/secrets/secret_api_key.txt` is a simple text file that might look like this
+
+```txt
+R^Y$qzKzSJw51K9zP$pQ3R3N
+```
+
+Equivalently, you can pipe the value to docker via stdin like this
+
+```sh
+echo "R^Y$qzKzSJw51K9zP$pQ3R3N" | docker secret create secret_api_key -
+```
+
+Now, with the secret defined, we can deploy the function like this
+
+
+```sh
+$ echo "R^Y$qzKzSJw51K9zP$pQ3R3N" | docker secret create secret_api_key -
+$ faas-cli deploy -f ./samples.yml --secret secret_api_key
+$ curl -H "Content-Type: application/json" \
+     -X POST \
+     -H "X-Api-Key: R^Y$qzKzSJw51K9zP$pQ3R3N" \
+     -d '{}' \
+     http://localhost:8080/function/protectedapi
+
+Unlocked the function!
+
+$ curl -H "Content-Type: application/json" \
+     -X POST \
+     -H "X-Api-Key: wrong_key" \
+     -d '{}' \
+     http://localhost:8080/function/protectedapi
+
+Access denied!
```

Comment:
Something a bit unexpected with the ApiKeyProtected example, its letting a sizeable substring match the key.
```
$ curl -H "Content-Type: application/json" -X POST -H "X-Api-Key: R^Y$qzKzSJw51K9zP$pQ3R3N" -d '{}' http://localhost:8080/function/protectedapi
Unlocked the function!

$ curl -H "Content-Type: application/json" -X POST -H "X-Api-Key: R^Y$q" -d '{}' http://localhost:8080/function/protectedapi
Unlocked the function!

$ curl -H "Content-Type: application/json" -X POST -H "X-Api-Key: R^Y$" -d '{}' http://localhost:8080/function/protectedapi
Access denied!

$ curl -H "Content-Type: application/json" -X POST -H "X-Api-Key: blahblahblah" -d '{}' http://localhost:8080/function/protectedapi
Access denied!
```
Are you seeing the same?

Reply:
I was able to reproduce using your curl command as well
```
➜  curl -H "Content-Type: application/json" -X POST -H "X-Api-Key: R^Y$q" -d '{}' http://localhost:8080/function/protectedapi
Unlocked the function!
```

That is really weird, the handler is fairly simple

```go
func handle(body []byte) {
	key := os.Getenv("Http_X_Api_Key")

	secretBytes, err := ioutil.ReadFile("/run/secrets/secret_api_key")
	if err != nil {
		log.Fatal(err)
	}

	secret := strings.TrimSpace(string(secretBytes))

	if key == secret {
		fmt.Println("Unlocked the function!")
	} else {
		fmt.Println("Access denied!")
	}
}
```

it seems like `golang` is doing some sort of regex match instead

Reply:
Odd, if you test the string the go playground, it works as expected https://play.golang.org/p/EjWkzE1wXd

Reply:
This example includes the byte to string conversion and the trim: https://play.golang.org/p/yCHDYNPlkC but it still works as expected

Reply:
Just found the "bug", the `$` is trying to do variable substitution in bash, so the secret actually looks like `R^Y` inside the container and the key in the curl command `R^Y$q` also looks like `R^Y`

Reply:
So how do we fix that?

Reply:
I don't think we can "fix" it because it is the terminal doing the variable substitution,  by the time the value is accessible to any part of `faas` or `docker`, it has already been modified.   I updated the walkthrough already to remove the `$` that causes the issue.  I think the only other thing we could do is warn people about using `$` in a secret when creating the secret using `echo` in the terminal.  Basically, don't use `$` if you plan on expressing the value in the terminal _or_ ensure that you escape the `$` correctly.

Reply:
OK 👍